### PR TITLE
Fix structured return from key signing in a number of negative cases

### PR DESCRIPTION
**Pull request by tdrl**
Also: Refactor for cleanliness.  Add tests.  Fix type definitions.  Fix type checking.

Fixes: #6875

### CODE REVIEW

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 1]
```diff
@@ -860,31 +874,32 @@ export class AmpA4A extends AMP.BaseElement {
...
+                      `error importing keys for service: ${jwkSet.serviceName}`,
+                      err, this.element);
+                  return null;
+                })),
+          };
         }).catch(err => {
           user().error(
               TAG, this.element.getAttribute('type'), err, this.element);
-          return {serviceName: currServiceName};
+          return {serviceName: currServiceName, keys: []};
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Given we are fetching these before we need them, let's add a TODO here that we could re-attempt to fetch the keys against if it fails

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 2]
```diff
@@ -1150,6 +1150,236 @@ describe('amp-a4a', () => {
...
+      return createAdTestingIframePromise().then(f => {
+        fixture = f;
+        win = fixture.win;
+        a4aElement = createA4aElement(fixture.doc);
+        return fixture;
+      });
+    });
+
+    function verifyIsKeyInfo(keyInfo) {
+      expect(keyInfo).not.to.be.null;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: expect(keyInfo).to.be.ok;

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 3]
```diff
@@ -1150,6 +1150,236 @@ describe('amp-a4a', () => {
...
+      expect(keyInfo).to.have.all.keys(
+          ['serviceName', 'hash', 'cryptoKey']);
+      expect(keyInfo.serviceName).to.be.a('string').and.not.to.equal('');
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: I didn't realize you could and... is that common practice as opposed to a separate expect item?

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 4]
```diff
@@ -1150,6 +1150,236 @@ describe('amp-a4a', () => {
...
+          }).returns(Promise.reject(
+              new TypeError('some random network error')));
+      expect(win.ampA4aValidationKeys).not.to.exist;
+      // Key fetch happens on A4A class construction.
+      const unusedA4a = new MockA4AImpl(a4aElement);  // eslint-disable-line no-unused-vars
+      const result = win.ampA4aValidationKeys;
+      expect(result).to.be.instanceof(Array);
+      expect(result).to.have.lengthOf(1);  // Only one service.
+      return result[0].then(serviceInfo => {
+        console.log('serviceInfo[keys] = ', serviceInfo['keys']);
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: Remove

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 5]
```diff
@@ -115,6 +115,12 @@ export let AdResponseDef;
     }} */
 let CreativeMetaDataDef;
 
+/** @typedef {{serviceName: string, keys: !Array<!Promise<?PublicKeyInfoDef>>}} */
+let SingleServiceKeySetInfoDef;
+
+/** @typedef {Array<!Promise<!SingleServiceKeySetInfoDef>>} */
+let PerServiceKeySetsInfoDef;
```
Comment (tdrl) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Yeah, not my best naming job ever, I admit.  :-(  Tried to revise and also add comments to clarify.  Let me know if you still have concerns.

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 6]
```diff
@@ -224,9 +230,11 @@ export class AmpA4A extends AMP.BaseElement {
     /** @private {boolean} whether creative has been verified as AMP */
     this.isVerifiedAmpCreative_ = false;
 
-    /** @private {Array<!Promise<!Array<!Promise<?PublicKeyInfoDef>>>>} */
-    this.win.ampA4aValidationKeys =
-      this.win.ampA4aValidationKeys || this.getKeyInfoSets_();
+    if (!this.win.ampA4aValidationKeys) {
+      /** @type {!PerServiceKeySetsInfoDef} */
+      const forTypeSafety = this.getKeyInfoSets_();
```
Comment (tdrl) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Done.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 1 above)
Comment (tdrl) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Done.

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 2 above)
Comment (tdrl) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: Done.

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 3 above)
Comment (tdrl) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: You can sometimes.  It's in the Chai API, anyway.  Not clear how often it's done.  Just felt slightly clearer to me than successive calls, but I'm not strongly attached -- would you prefer it as separate `expect()`s?

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 4 above)
Comment (tdrl) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: Oops.  Forgot that one.  Thanks for catching it.  Done.

**Comment by tdrl** [comment]
@muxin Okay, Keith is happy and I'm good.  Please take a look and, if it looks good to you, merge this?

**Pull request merged**